domain/relation/state/relation_test.go

```diff
@@ -24,17 +25,491 @@ import (
 	"github.com/juju/juju/internal/uuid"
 )
 
-type relationSuite struct {
-	schematesting.ModelSuite
+// addRelationSuite is a test suite dedicated to check functionalities
+// related to adding relation.
+// It extends baseRelationSuite to leverage common setup and utility methods
+// for relation-related testing and provides more builder dedicated for this
+// specific context.
+type addRelationSuite struct {
+	baseRelationSuite
+
+	// charmByApp maps application IDs to their associated charm IDs for quick
+	// lookup during tests.
+	charmByApp map[coreapplication.ID]corecharm.ID
+}
+
+var _ = gc.Suite(&addRelationSuite{})
+
+func (s *addRelationSuite) SetUpTest(c *gc.C) {
+	s.baseRelationSuite.SetUpTest(c)
+	s.charmByApp = make(map[coreapplication.ID]corecharm.ID)
+}
+
+func (s *addRelationSuite) TestAddRelation(c *gc.C) {
+	// Arrange
+	relProvider := charm.Relation{
+		Name:  "prov",
+		Role:  charm.RoleProvider,
+		Scope: charm.ScopeGlobal,
+	}
+	relRequirer := charm.Relation{
+		Name:  "req",
+		Role:  charm.RoleRequirer,
+		Scope: charm.ScopeGlobal,
+	}
+	appUUID1 := s.addApplication(c, "application-1")
+	appUUID2 := s.addApplication(c, "application-2")
+	epUUID1 := s.addApplicationEndpointFromRelation(c, appUUID1, relProvider)
+	epUUID2 := s.addApplicationEndpointFromRelation(c, appUUID2, relRequirer)
+	epUUID3 := s.addApplicationEndpointFromRelation(c, appUUID2, relProvider)
+	epUUID4 := s.addApplicationEndpointFromRelation(c, appUUID1, relRequirer)
+
+	// Act
+	ep1, ep2, err := s.state.AddRelation(context.Background(), relation.CandidateEndpointIdentifier{
+		ApplicationName: "application-1",
+		EndpointName:    "prov",
+	}, relation.CandidateEndpointIdentifier{
+		ApplicationName: "application-2",
+		EndpointName:    "req",
+	})
+	c.Assert(err, jc.ErrorIsNil, gc.Commentf("(Act) unexpected error while inserting the first relation: %s",
+		errors.ErrorStack(err)))
+	ep3, ep4, err := s.state.AddRelation(context.Background(), relation.CandidateEndpointIdentifier{
+		ApplicationName: "application-1",
+		EndpointName:    "req",
+	}, relation.CandidateEndpointIdentifier{
+		ApplicationName: "application-2",
+		EndpointName:    "prov",
+	})
+	c.Assert(err, jc.ErrorIsNil, gc.Commentf("(Act) unexpected error while inserting the second relation: %s",
+		errors.ErrorStack(err)))
+
+	// Assert
+	c.Check(ep1, gc.Equals, relation.Endpoint{
+		ApplicationName: "application-1",
+		Relation:        relProvider,
+	})
+	c.Check(ep2, gc.Equals, relation.Endpoint{
+		ApplicationName: "application-2",
+		Relation:        relRequirer,
+	})
+	c.Check(ep3, gc.Equals, relation.Endpoint{
+		ApplicationName: "application-1",
+		Relation:        relRequirer,
+	})
+	c.Check(ep4, gc.Equals, relation.Endpoint{
+		ApplicationName: "application-2",
+		Relation:        relProvider,
+	})
+	epUUIDsByRelID := make(map[int][]string)
+	err = s.TxnRunner().StdTxn(context.Background(), func(ctx context.Context, tx *sql.Tx) error {
```

Comment:
todo: This test is too long, move the call to `s.TxnRunner().StdTxn` to a new helper method.